core/src/transfer/data.rs

```diff
@@ -39,6 +43,126 @@ pub enum TransactionData {
+            TransactionData::Memo(m) => {
+                bytes.extend(m);
+            }
+            TransactionData::Blob(blobs) => {
+                // Hash the blobs and extend the signature_message with the
+                // result, using the same serialization format used by
+                // TransactionData::Memo, so that the legacy transfer contract
+                // can still verify it after being converted with `blob_to_memo`
+                let mut hasher = Sha256::new();
+                for blob in blobs {
+                    hasher.update(blob.to_hash_input_bytes());
+                }
+                bytes.extend(hasher.finalize());
+            }
+        }
+
+        bytes
+    }
+
+    /// Serialize a `TransactionData` into a variable length byte buffer.
+    #[must_use]
+    pub fn to_var_bytes(&self) -> Vec<u8> {
+        let mut bytes = Vec::new();
+        // serialize the contract call, deployment or memo, if present.
+        match &self {
+            TransactionData::Call(call) => {
+                bytes.push(1);
+                bytes.extend(call.to_var_bytes());
+            }
+            TransactionData::Deploy(deploy) => {
+                bytes.push(2);
+                bytes.extend(deploy.to_var_bytes());
+            }
+            TransactionData::Memo(memo) => {
+                bytes.push(3);
+                bytes.extend((memo.len() as u64).to_bytes());
+                bytes.extend(memo);
+            }
+            TransactionData::Blob(blobs) => {
+                bytes.push(4);
+                // Maybe we can use u8 as length
+                bytes.extend((blobs.len() as u64).to_bytes());
+                for blob in blobs {
+                    bytes.extend(blob.to_var_bytes());
+                }
+            }
+        }
+
+        bytes
+    }
+
+    /// Deserialize the optional `TransactionData` from bytes slice.
+    ///
+    /// # Errors
+    /// Errors when the bytes are not canonical.
+    pub fn from_slice(buf: &[u8]) -> Result<Option<Self>, BytesError> {
+        let mut buf = buf;
+
+        // deserialize contract call, deploy data, or memo, if present
+        let data = match u8::from_reader(&mut buf)? {
+            0 => None,
+            1 => Some(TransactionData::Call(ContractCall::from_slice(buf)?)),
+            2 => {
+                Some(TransactionData::Deploy(ContractDeploy::from_slice(buf)?))
+            }
+            3 => {
+                // we only build for 64-bit so this truncation is impossible
+                #[allow(clippy::cast_possible_truncation)]
+                let size = u64::from_reader(&mut buf)? as usize;
+
+                if buf.len() != size || size > MAX_MEMO_SIZE {
+                    return Err(BytesError::InvalidData);
+                }
+
+                let memo = buf[..size].to_vec();
+                Some(TransactionData::Memo(memo))
+            }
+            4 => {
+                let blobs_len = u8::from_reader(&mut buf)?;
```

Comment:
shouldn't this match `to_var_bytes`? we have `u8` here, and `u64` there

Reply:
nice catch